### ISSUE TITLE
Publish files and hashes of Windows installers on GitHub

### ISSUE THREAD

**Issue by rugk** [qTox/qTox#2670]
IMO it would be nice if you could publish the Windows installers on GitHub [in the releases part](https://github.com/tux3/qTox/releases). This would allow users to easily switch back to any version they like if they e.g. encounter problems.
To further improve the security I'd suggest you to also add the hashes of these files in the description there. (Recommend hashes: SHA-265 and SHA-512)

**Comment by tux3** [comment]
We used to publish releases on GitHub but it was a bit of a mess after having hundreds of them, and it was taking way too much space on GitHub's servers.
I think we're going to be staying on Jenkins for the time beeing. For switching to previous versions, we could keep older versions of the installer on Jenkins depending how much disk space is left.

**Comment by rugk** [comment]
> but it was a bit of a mess after having hundreds of them

Well... now you are in stable and you should not have so many releases anymore.
I don't know how you did it before, because previously you had no version numbers, but I think the releases were much more in the past. Did you even published a release for every commit?!

> and it was taking way too much space on GitHub's servers.

Where exactly did you published them? In a repo?
If so this is obviously the wrong way to do it. Just add them as a binary file [when you create a release](https://help.github.com/articles/creating-releases/).
GitHub uses Amazon AWS servers for this, so they have plenty of space.

> I think we're going to be staying on Jenkins for the time beeing.

I don't say you should move away from Jenkins and the build links there. Just add the releases to GitHub additionally.

Please reopen this issue.

**Comment by tux3** [comment]
> Well... now you are in stable and you should not have so many releases anymore.
> I don't know how you did it before, because previously you had no version numbers, but I think the releases were much more in the past. Did you even published a release for every commit?!

Pretty much, yes.

> Where exactly did you published them? In a repo?
> If so this is obviously the wrong way to do it. Just add them as a binary file when you create a release.
> GitHub uses Amazon AWS servers for this, so they have plenty of space.

We obviously weren't pushing binaries to the repo, or nobody would even be able to clone it at this point...
But the problem is that it's easy to keep adding releases automatically, but there's no automated way to prune old ones, so at 2*30MB just for the Windows binaries, this adds up quickly.

**Comment by rugk** [comment]
> Pretty much, yes.

And there are pretty less now, so the amount of releases is certainly no problem.

> But the problem is that it's easy to keep adding releases automatically, but there's no automated way to prune old ones, so at 2*30MB just for the Windows binaries, this adds up quickly.

But the releases are on GitHubs AWS server, so why even bother?
This is not your but GitHubs problem. So you do not need to delete old releases anyway. If GitHub has a problem they'll notify you. (But I bet that's not going to happen)

A current release is 20MB and GitHubs [only limitation](https://help.github.com/articles/about-releases/#limitations-on-binary-files) is that _one single file_ must not be larger than 2GB and you're far away from reaching this:

> We don't limit the total size of your binary release files, nor the bandwidth used to deliver them. However, each individual file must be under 2GB in size.

And if GitHubs sets such a limitation I'm sure there are users who upload files up-to 2GB there, so with a mere 20MB you're not going to have any problems at all.

**Comment by sudden6** [comment]
The problem arises when others want to clone the repository to work on it. Because of how git works, every file that ever existed in the repository is saved. Nobody wants to download GBs of outdated binaries just to make some small change to the code.

**Comment by rugk** [comment]
The binaries are cloned? Really?
I don't think GitHub stores them in the repo and clones them if they are there.

Can you show me a source for this?

**Comment by sudden6** [comment]
@rugk sorry I was wrong, I missed your part about Github Releases and thought you wanted to store the binaries inside the repo.

**Comment by rugk** [comment]
Yes, this is the important part. And as for GitHub releases you can upload very huge files and it does not matter in any way. So @tux3: Do you finally want to do this?

**Comment by tux3** [comment]
@rugk 
It sounds doable, but there's a fragmentation problem, because if I upload only Windows installers here, Linux users that find the page could think we don't have Linux releases (the Linux releases are on Jenkins/tox.chat with all the other releases). 
And if I upload Linux packages for each release too, there's 30 different combinations of packages that I would need to upload plus instructions to install the repo for later updates, that's just not practical.

Maybe I should make a Linux installer that I can upload with the Windows installer. 
Or maybe a qTox site where you can download all the releases and hashes with a nice UI for each OS.

**Comment by rugk** [comment]
I think Linux users know that they mostly have to do a way with their package managers and additionally in the Readme and the wiki are enough "hints", links and descriptions which make clear that you can also use this on GitHub.
AFAIK the best thing a Linux "installer" can do is adding the apt sources and fetch them. I don't think you really need a separate installer for this as all Linux users should be experienced enough to do this on their own.

Basically you can just add the Windows installer binaries, publish the hashes of these and this should be enough.
If you think it helps you can also publish the links to the Linux binaries (which represent exactly the published version of course) and their hashes.

**Comment by tux3** [comment]
> I think Linux users know that they mostly have to do a way with their package managers and additionally in the Readme and the wiki are enough "hints", links and descriptions which make clear that you can also use this on GitHub.
> AFAIK the best thing a Linux "installer" can do is adding the apt sources and fetch them. I don't think you really need a separate installer for this as all Linux users should be experienced enough to do this on their own.

The thing is that I've had several people using Ubuntu, Elementary OS, or other Linux distributions that had trouble installing the repositories, it's already pretty hard to install Linux on qTox if you're not an expert. 
Having different pages that won't work for everyone will only add to the confusion imo.
If we're going to publish older releases and installers, we might as well add them to the existing page on tox.chat and Jenkins, or make a new page that has everything everyone needs.

> Basically you can just add the Windows installer binaries, publish the hashes of these and this should be enough.

That's definitely possible, I'd need to write a script and find a trusted place to run it (since the script needs full access to the repo).

> If you think it helps you can also publish the links to the Linux binaries (which represent exactly the published version of course) and their hashes.

It's more that I don't want to have different pages that each have different parts of the information. I'd rather have one place where people know they can go to install qTox.

**Comment by rugk** [comment]
[**@tux3**](https://github.com/tux3) commented on [31. Dez. 2015, 15:22 MEZ](https://github.com/tux3/qTox/issues/2670#issuecomment-168202939):

> The thing is that I've had several people using Ubuntu, Elementary OS, or other Linux distributions that had trouble installing the repositories, it's already pretty hard to install Linux on qTox if you're not an expert.

I would install it the other way round, but if you think so... :wink: 

> If we're going to publish older releases and installers, we might as well add them to the existing page on tox.chat and Jenkins, or make a new page that has everything everyone needs.

This would also be possible. Like an "archive" of releases.

> That's definitely possible, I'd need to write a script and find a trusted place to run it (since the script needs full access to the repo).

Can't you just do it manually? I mean you update the links to the releases and sign them manually too AFAIK and this is not something which happens so often.

> It's more that I don't want to have different pages that each have different parts of the information. I'd rather have one place where people know they can go to install qTox.

Okay. And if you have a site with links for linux builds and the "download for windows" part links to the GitHub releases?

**Comment by rugk** [comment]
As I see you now have Linux installers too. So what about publishing hashes now? (at least of the installers)
